File: src/main/java/com/stal111/forbidden_arcanus/data/recipes/ApplyModifierRecipeProvider.java
package com.stal111.forbidden_arcanus.data.recipes;

import com.stal111.forbidden_arcanus.common.item.modifier.ItemModifier;
import com.stal111.forbidden_arcanus.core.init.ModItems;
import com.stal111.forbidden_arcanus.core.init.other.ModItemModifiers;
import com.stal111.forbidden_arcanus.data.recipes.builder.ApplyModifierRecipeBuilder;
import net.minecraft.core.Holder;
import net.minecraft.core.HolderLookup;
import net.minecraft.world.level.ItemLike;
import net.valhelsia.valhelsia_core.datagen.recipes.RecipeSubProvider;
import net.valhelsia.valhelsia_core.datagen.recipes.ValhelsiaRecipeProvider;

/**
 * @author stal111
 * @since 2022-10-21
 */
public class ApplyModifierRecipeProvider extends RecipeSubProvider {


    public ApplyModifierRecipeProvider(ValhelsiaRecipeProvider provider) {
        super(provider);
    }

    @Override
    protected void registerRecipes(HolderLookup.Provider lookupProvider) {
        this.modifier(ModItems.ETERNAL_STELLA.get(), ModItemModifiers.ETERNAL);
        this.modifier(ModItems.SMELTER_PRISM.get(), ModItemModifiers.FIERY);
        this.modifier(ModItems.FERROGNETIC_MIXTURE.get(), ModItemModifiers.MAGNETIZED);
        this.modifier(ModItems.TERRASTOMP_PRISM.get(), ModItemModifiers.DEMOLISHING);
        this.modifier(ModItems.SEA_PRISM.get(), ModItemModifiers.AQUATIC);
    }

    private void modifier(ItemLike addition, Holder<ItemModifier> modifier) {
        this.add(ApplyModifierRecipeBuilder.of(ModItems.DARKSTONE_UPGRADE_SMITHING_TEMPLATE.get(), addition, modifier));
    }
}

File: datagen/src/main/kotlin/com/stal111/forbidden_arcanus/datagen/recipes/ApplyModifierRecipeProvider.kt
package com.stal111.forbidden_arcanus.datagen.recipes

import com.stal111.forbidden_arcanus.common.item.modifier.BuiltInItemModifiers
import com.stal111.forbidden_arcanus.common.item.modifier.ItemModifier
import com.stal111.forbidden_arcanus.core.init.ModItems
import com.stal111.forbidden_arcanus.data.recipes.builder.ApplyModifierRecipeBuilder
import net.minecraft.core.Holder
import net.minecraft.core.HolderLookup
import net.minecraft.data.recipes.RecipeOutput
import net.minecraft.world.level.ItemLike
import net.valhelsia.dataforge.recipe.RecipeSubProvider

class ApplyModifierRecipeProvider(provider: HolderLookup.Provider, recipeOutput: RecipeOutput) :
    RecipeSubProvider(provider, recipeOutput) {
    override fun buildRecipes() {
        this.modifier(ModItems.ETERNAL_STELLA.get(), registries.holderOrThrow(BuiltInItemModifiers.ETERNAL))
        this.modifier(ModItems.SMELTER_PRISM.get(), registries.holderOrThrow(BuiltInItemModifiers.FIERY))
        this.modifier(ModItems.FERROGNETIC_MIXTURE.get(), registries.holderOrThrow(BuiltInItemModifiers.MAGNETIZED))
        this.modifier(ModItems.TERRASTOMP_PRISM.get(), registries.holderOrThrow(BuiltInItemModifiers.DEMOLISHING))
        this.modifier(ModItems.SEA_PRISM.get(), registries.holderOrThrow(BuiltInItemModifiers.AQUATIC))
        this.modifier(ModItems.SOUL_BINDING_CRYSTAL.get(), registries.holderOrThrow(BuiltInItemModifiers.SOULBOUND))
    }

    private fun modifier(addition: ItemLike, modifier: Holder<ItemModifier>) {
        this.add(ApplyModifierRecipeBuilder.of(ModItems.APPLY_MODIFIER_SMITHING_TEMPLATE.get(), addition, modifier), "apply_modifier/" + modifier.key!!.location().path)
    }
}
